README.md
# Sudoku

Another clone of the Japanese puzzle

Made by Vegard Løkken (2012)

Feel free to share this game


CHANGELOG
*** VERSION 0.10
- Initial beta release

*** VERSION 0.15
- Fixed hanging loop on createGame
- Added retry possibility

*** VERSION 0.20
- Added status bar
- Fixed retry bug that happened after finishGame
- Fixed finishGame bug by using gameOriginal instead of game

*** VERSION 0.25
- Added highscore list


src/no/vtek/sudoku/StatusBar.java
/*
 * The MIT License
 *
 * Copyright 2015 Vegard Løkken
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package no.vtek.sudoku;

import java.awt.BorderLayout;
import java.awt.Color;
import javax.swing.JLabel;
import javax.swing.JPanel;

/**
 * The sudoku game's status bar
 *
 * @author Vegard Løkken
 * @version 0.25
 */
class StatusBar extends JPanel {

	private SudokuGame frame;
	private JLabel statusText = new JLabel("Welcome to Sudoku");
	private JLabel time = new JLabel("Time: 0 ");
	private TimeThread thread = null;

	public StatusBar(SudokuGame frame) {
		this.frame = frame;

		setLayout(new BorderLayout());

		statusText.setForeground(Color.gray);
		time.setForeground(Color.gray);

		add(statusText, BorderLayout.WEST);
		add(time, BorderLayout.EAST);
	}

	public void setStatus(String status) {
		statusText.setText(status);
	}

	public void startTime() {
		if (thread != null && thread.running) stopTime();
		
		thread = new TimeThread();
		thread.start();
	}

	public void stopTime() {
		thread.running = false;
	}

	public void pauseTime() {
		thread.pause = true;
	}

	public void resumeTime() {
		thread.pause = false;
	}

	public int getTime() {
		return thread.timeCount;
	}

	private class TimeThread extends Thread {
		private boolean running = true;
		private boolean pause = false;
		private int timeCount = 0;
		private long last;

		@Override
		public void run() {
			last = System.currentTimeMillis();
			
			while (running) {
				long now = System.currentTimeMillis();
				long diff = now - last;
				
				last = now;
					
				if (!pause) {
					timeCount += diff;
					time.setText("Time: " + timeCount/1000 + " ");
				}

				try {
					Thread.sleep(50);
				} catch (InterruptedException ex) {
				}
			}
		}
	}
}


src/no/vtek/sudoku/SudokuBoard.java
/*
 * The MIT License
 *
 * Copyright 2015 Vegard Løkken
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package no.vtek.sudoku;

import java.awt.Color;
import java.awt.Dimension;
import java.awt.FlowLayout;
import java.awt.Font;
import java.awt.event.MouseEvent;
import java.awt.event.MouseListener;
import java.util.ArrayList;
import javax.swing.*;

/**
 * The sudoku game board
 * 
 * @author Vegard Løkken
 * @version 0.25
 */
public class SudokuBoard extends JPanel implements MouseListener {
	private SudokuGame frame;
	private int[][] game;
	private int[][] gameOriginal;
	private boolean[][] networkCheck;
	private JLabel[][] tiles = new JLabel[9][9];
	private int chosenNumber = 0;
	private NumberChooser numberChooserDialog = new NumberChooser();
	JLabel activeTile = null;

	public SudokuBoard(SudokuGame frame, int[][] game) {
		this.frame = frame;
		this.game = game;
		
		/* Making a copy of the original so that we can remember which
		 * fields are editable */
		gameOriginal = new int[game.length][game[0].length];
		for (int i=0; i<game.length; i++) {
			for (int j=0; j<game[0].length; j++) {
				gameOriginal[i][j] = game[i][j];
			}
		}
		
		networkCheck = new boolean[game.length][game[0].length];
		
		setPreferredSize(new Dimension(500, 500));
		setLayout(null);
		
		for (int i=0; i<game.length; i++) {
			for (int j=0; j<game[0].length; j++) {
				tiles[i][j] = new JLabel();
				tiles[i][j].setBackground(Color.white);
				tiles[i][j].setOpaque(true);
				tiles[i][j].addMouseListener(this);
				tiles[i][j].setFont(new Font("Arial", Font.BOLD, 18));
				
				if (game[i][j] != 0) {
					tiles[i][j].setText(new Integer(game[i][j]).toString());
				} else {
					//tiles[i][j].setText("?");
					tiles[i][j].setForeground(Color.red);
					//tiles[i][j].setToolTipText("Click to change");
				}
				
				tiles[i][j].setHorizontalAlignment(SwingConstants.CENTER);
				tiles[i][j].setBorder(
						BorderFactory.createLineBorder(Color.black));
				tiles[i][j].setBounds(j*(50+5)+5, i*(50+5)+5, 50, 50);
				add(tiles[i][j]);
			}
		}
		
		JPanel squareFrame;
		for (int i=2; i<497; i+=165) {
			for (int j=2; j<497; j+=165) {
				squareFrame = new JPanel();
				squareFrame.setBounds(j, i, 166, 166);
				squareFrame.setBorder(BorderFactory.createLineBorder(Color.black, 2));
				add(squareFrame);
			}
		}
		
	}
	
	private int[] findIndex(JLabel button) {
		int[] ret = new int[]{-1, -1};
		
		for (int i=0; i<tiles.length; i++) {
			for (int j=0; j<tiles[0].length; j++) {
				if (tiles[i][j] == button) {
					ret = new int[]{j, i};
				}
			}
		}
		
		return ret;
	}
	
	private boolean boardCompleted() {
		boolean completed = true;
		
		outer:
		for (int i=0; i<game.length; i++) {
			for (int j=0; j<game[0].length; j++) {
				if (game[i][j] == 0) {
					completed = false;
					break outer;
				}
			}
		}
		
		return completed;
	}

	@Override
	public void mouseClicked(MouseEvent e) {
		if (frame.inGame && e.getSource() instanceof JLabel) {
			int[] index = findIndex((JLabel)e.getSource());

			int x = index[0];
			int y = index[1];

			int clickedNumber = frame.redo ? gameOriginal[y][x] : game[y][x];

			if (clickedNumber == 0) {
				
					numberChooserDialog.showDialog(x, y);
					
					setNumber(x, y, chosenNumber);
			}
		}
	}

	@Override
	public void mousePressed(MouseEvent e) {}

	@Override
	public void mouseReleased(MouseEvent e) {}

	@Override
	public void mouseEntered(MouseEvent e) {
		if (frame.inGame && e.getSource() instanceof JLabel) {
			
			// Color those who are affected
			if (frame.lineAssistance) {
				activeTile = (JLabel) e.getSource();

				for (JLabel tile : getNeighbours(activeTile)) {
					tile.setBackground(new Color(200, 225, 250));
				}

				activeTile.setBackground(new Color(100, 150, 225));
			}
			
			
			int[] index = findIndex((JLabel)e.getSource());
			int x = index[0];
			int y = index[1];
			if (gameOriginal[y][x] == 0) {
				frame.statusBar.setStatus("Click to change");
			}
		}
	}

	@Override
	public void mouseExited(MouseEvent e) {
		
		for (int i=0; i<tiles.length; i++) {
			for (int j=0; j<tiles[0].length; j++) {
			tiles[i][j].setBackground(Color.white);
			}
		}
		
		frame.statusBar.setStatus(null);
	}
	
	private void setNumber(int x, int y, int num) {
		game[y][x] = num;

		if (num > 0) {

			tiles[y][x].setText(new Integer(num).toString());

			/* Check for win */
			if (boardCompleted()) {
				if (SudokuFactory.checkValid(game)) {
					frame.statusBar.stopTime();
					frame.gameOver();

				} else {
					JOptionPane.showMessageDialog(frame,
						"You have completed the board, but it " +
						"doesn't seem to fulfill sudoku rules. " +
						"Take one more look at it and try again.");
				}
			}
		} else {

			tiles[y][x].setText(null);
		}
	}
	
	private ArrayList<JLabel> getNeighbours(JLabel activeTile) {
		ArrayList<JLabel> neighbours = new ArrayList<JLabel>();
		
		int[] index = findIndex(activeTile);
		int x = index[0], y = index[1];
		
		// Rows
		for (int i=0; i<tiles.length; i++) {
			neighbours.add(tiles[i][x]);
		}
		
		// Columns
		for (int i=0; i<tiles[0].length; i++) {
			neighbours.add(tiles[y][i]);
		}
		
		// Square
		int startCol = (x / 3) * 3;
		int startRow = (y / 3) * 3;

		for (int i=startRow; i<startRow+3; i++) {
			for (int j=startCol; j<startCol+3; j++) {
				neighbours.add(tiles[i][j]);
			}
		}
		
		return neighbours;
	}
	
	public void finishGame() {
		
		int[][] finished = SudokuFactory.finishGame(gameOriginal);
		
		if (finished != null) {
			frame.inGame = false;
			frame.statusBar.stopTime();
			
			game = finished;

			for (int i=0; i<tiles.length; i++) {
				for (int j=0; j<tiles[0].length; j++) {
					tiles[i][j].setText(new Integer(game[i][j]).toString());
				}
			}
			
		} else {
			System.out.println("ERROR: Couldn't finish game.");
		}
	}
	
	public void retryGame() {
		
		game = new int[gameOriginal.length][gameOriginal[0].length];
		for (int i=0; i<gameOriginal.length; i++) {
			for (int j=0; j<gameOriginal[0].length; j++) {
				game[i][j] = gameOriginal[i][j];
				
				if (game[i][j] > 0) {
					tiles[i][j].setText(
						new Integer(gameOriginal[i][j]).toString());
				} else {
					tiles[i][j].setText(null);
				}
			}
		}
		
		frame.inGame = true;
		frame.statusBar.startTime();
		
	}
	
	private class NumberChooser extends JDialog implements MouseListener {
		
		private ArrayList<JLabel> buttons = new ArrayList<JLabel>();
		private JLabel xButton;
		
		public NumberChooser() {
			super(frame, true);
			setUndecorated(true);
			setResizable(false);
			setLayout(new FlowLayout(FlowLayout.CENTER, 5, 5));
			
			xButton = new JLabel("clear");
			xButton.addMouseListener(this);
			xButton.setPreferredSize(new Dimension(100, 15));
			xButton.setBorder(BorderFactory.createLineBorder(Color.black));
			xButton.setBackground(Color.black);
			xButton.setForeground(Color.white);
			xButton.setOpaque(true);
			xButton.setHorizontalAlignment(SwingConstants.CENTER);
			xButton.setFont(new Font("Arial", Font.BOLD, 14));
			add(xButton);
			
			for (int i=1; i<10; i++) {
				JLabel button = new JLabel(""+i);
				button.addMouseListener(this);
				button.setPreferredSize(new Dimension(30, 30));
				button.setBorder(BorderFactory.createLineBorder(Color.black));
				button.setBackground(Color.white);
				button.setOpaque(true);
				button.setHorizontalAlignment(SwingConstants.CENTER);
				button.setFont(new Font("Arial", Font.BOLD, 14));
				add(button);
				
				buttons.add(button);
			}
			
		}
		
		public void showDialog(int x, int y) {

			setBounds(frame.getBounds().x + (x*55),
				frame.getBounds().y + ((y+1)*55), 110, 130);
			
			changeButtonBackground(frame.numberAssistance, x, y);
			
			setVisible(true);
		}
		
		private void changeButtonBackground(boolean help, int x, int y) {
			if (help) {
				int[] validNumbers = SudokuFactory.getPossibleNumbers(
					game, y, x);
				
				for (JLabel button : buttons) {
					int num = Integer.parseInt(button.getText());
					if (validNumbers[num] == 1) { // Not valid
						button.setBackground(Color.lightGray);
					} else {
						button.setBackground(Color.white);
					}
				}
			} else {
				for(JLabel button : buttons) {
					button.setBackground(Color.white);
				}
			}
		}

		@Override
		public void mouseClicked(MouseEvent e) {
			JLabel src = (JLabel) e.getSource();

			if (src == xButton) {
				chosenNumber = 0;
			} else {
				chosenNumber = Integer.parseInt(src.getText());
			}

			setVisible(false);
		}

		@Override
		public void mousePressed(MouseEvent e) {}

		@Override
		public void mouseReleased(MouseEvent e) {}

		@Override
		public void mouseEntered(MouseEvent e) {}

		@Override
		public void mouseExited(MouseEvent e) {}
		
	}
}


src/no/vtek/sudoku/SudokuDialog.java
/*
 * The MIT License
 *
 * Copyright 2015 Vegard Løkken
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package no.vtek.sudoku;

import java.awt.BorderLayout;
import java.awt.FlowLayout;
import java.awt.Frame;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.util.ArrayList;
import javax.swing.JButton;
import javax.swing.JComponent;
import javax.swing.JDialog;
import javax.swing.JPanel;

/**
 * Generic modal dialog
 * 
 * @author Vegard Løkken
 * @version 0.25
 */
public class SudokuDialog extends JDialog implements ActionListener {
	private ArrayList<JComponent> comps;
	private JPanel compsPanel, buttonPanel;
	private JButton okButton, abortButton;
	private boolean ok = false;

	public SudokuDialog(Frame owner) {
		super(owner, true);
		setLayout(new BorderLayout());
		
		comps = new ArrayList<JComponent>();
		compsPanel = new JPanel();
		compsPanel.setLayout(new FlowLayout(FlowLayout.CENTER));
		
		buttonPanel = new JPanel();
		okButton = new JButton("OK");
		abortButton = new JButton("Abort");
		
		buttonPanel.add(okButton);
		buttonPanel.add(abortButton);
		
		okButton.addActionListener(this);
		abortButton.addActionListener(this);
		
	}
	
	public void addComp(JComponent comp) {
		comps.add(comp);
		compsPanel.add(comp);
	}
	
	public void open() {
		add(compsPanel, BorderLayout.NORTH);
		add(buttonPanel, BorderLayout.SOUTH);
		pack();
		this.setSize(getWidth()+50, getHeight());
		setVisible(true);
	}

	public ArrayList<JComponent> getComps() {
		return comps;
	}

	@Override
	public void actionPerformed(ActionEvent e) {
		Object source = e.getSource();
		
		if (source == okButton) {
			ok = true;
		}
		
		setVisible(false);
	}
	
}


src/no/vtek/sudoku/HighscoreItem.java
/*
 * The MIT License
 *
 * Copyright 2015 Vegard Løkken
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package no.vtek.sudoku;

import java.io.Serializable;
import java.util.Date;

/**
 * The highscore item contains name, difficulty level, time and date set
 * 
 * @author Vegard Løkken
 * @version 0.25
 */
class HighscoreItem implements Serializable {
	private String name;
	private int difficultyLevel;
	private int time;
	private Date date;

	public HighscoreItem(String name, int difficultyLevel, int time, Date date) {
		this.name = name;
		this.difficultyLevel = difficultyLevel;
		this.time = time;
		this.date = date;
	}

	public Date getDate() {
		return date;
	}

	public void setDate(Date date) {
		this.date = date;
	}

	public int getDifficultyLevel() {
		return difficultyLevel;
	}

	public void setDifficultyLevel(int difficultyLevel) {
		this.difficultyLevel = difficultyLevel;
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public int getTime() {
		return time;
	}

	public void setTime(int time) {
		this.time = time;
	}
	
}


src/no/vtek/sudoku/SudokuGame.java
/*
 * The MIT License
 *
 * Copyright 2015 Vegard Løkken
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package no.vtek.sudoku;

import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.ItemEvent;
import java.awt.event.ItemListener;
import javax.swing.*;

/**
 * Sudoku game
 *
 * @author Vegard Løkken
 * @version 0.25
 */
public class SudokuGame extends JFrame implements ActionListener, ItemListener {

	private MenuItem newGameItem;
	private MenuItem retryGameItem;
	private MenuItem highscoreGameItem;
	private MenuItem exitGameItem;
	private CheckboxMenuItem lineAssistanceMenuItem;
	private CheckboxMenuItem numberAssistanceMenuItem;
	private CheckboxMenuItem redoMenuItem;
	private MenuItem finishMenuItem;
	private MenuItem aboutMenuItem;
	private MenuItem helpMenuItem;
	protected SudokuBoard gamePanel;
	protected StatusBar statusBar;
	protected boolean inGame = false;
	protected boolean lineAssistance = true;
	protected boolean numberAssistance = false;
	protected boolean redo = true;
	protected int[][] game = null;
	private int difficultyLevel = -1;

	public SudokuGame() {
		super("Sudoku puzzle");
		setSize(500, 500);
		setDefaultCloseOperation(EXIT_ON_CLOSE);
		setResizable(false);

		// Try to use system look and feel
		try {
			UIManager.setLookAndFeel(
				UIManager.getSystemLookAndFeelClassName());
		} catch (Exception e) { }

		statusBar = new StatusBar(this);
		add(statusBar, BorderLayout.SOUTH);

		MenuBar menuBar = new MenuBar();
		setMenuBar(menuBar);

		Menu gameMenu = new Menu("Game");
		menuBar.add(gameMenu);

		newGameItem = new MenuItem("New game");
		newGameItem.addActionListener(this);

		retryGameItem = new MenuItem("Retry game");
		retryGameItem.addActionListener(this);

		highscoreGameItem = new MenuItem("Highscore list");
		highscoreGameItem.addActionListener(this);

		exitGameItem = new MenuItem("Exit");
		exitGameItem.addActionListener(this);

		gameMenu.add(newGameItem);
		gameMenu.add(retryGameItem);
		gameMenu.add(highscoreGameItem);
		gameMenu.add(exitGameItem);

		Menu optionsMenu = new Menu("Options");
		menuBar.add(optionsMenu);

		Menu assistance = new Menu("Assistance");
		optionsMenu.add(assistance);

		lineAssistanceMenuItem = new CheckboxMenuItem("Line Assistance");
		lineAssistanceMenuItem.addItemListener(this);
		lineAssistanceMenuItem.setState(true);

		numberAssistanceMenuItem = new CheckboxMenuItem("Number Assistance");
		numberAssistanceMenuItem.addItemListener(this);
		numberAssistanceMenuItem.setState(false);

		assistance.add(lineAssistanceMenuItem);
		assistance.add(numberAssistanceMenuItem);

		redoMenuItem = new CheckboxMenuItem("Redo possible");
		redoMenuItem.addItemListener(this);
		redoMenuItem.setState(true);

		optionsMenu.add(redoMenuItem);

		finishMenuItem = new MenuItem("Finish game");
		finishMenuItem.addActionListener(this);

		optionsMenu.add(finishMenuItem);

		Menu helpMenu = new Menu("Help");
		menuBar.add(helpMenu);

		aboutMenuItem = new MenuItem("About");
		aboutMenuItem.addActionListener(this);

		helpMenuItem = new MenuItem("Help");
		helpMenuItem.addActionListener(this);

		helpMenu.add(aboutMenuItem);
		helpMenu.add(helpMenuItem);

		setVisible(true);

		// Center
		this.setLocationRelativeTo(null);

		startNewGame();
	}

	protected void startNewGame() {

		boolean ok = true;

		if (inGame) {
			int ans = JOptionPane.showConfirmDialog(this,
							"A game is already in progress. "
							+ "Are you sure you want to start a new one?", "Abort",
							JOptionPane.YES_NO_OPTION);

			if (ans == JOptionPane.NO_OPTION) {
				ok = false;
			} else {
				remove(gamePanel);
				gamePanel = null;
				game = null;
			}
		} else if (gamePanel != null) {
			remove(gamePanel);
			gamePanel = null;
			game = null;
		}

		if (ok) {
			
			WelcomeDialog welcomeDialog = showWelcomeDialog();
			
			JButton dif = welcomeDialog.difficulty;
			if (dif == welcomeDialog.easyButton) {
				difficultyLevel = SudokuFactory.EASY;
			} else if (dif == welcomeDialog.mediumButton) {
				difficultyLevel = SudokuFactory.MEDIUM;
			} else if (dif == welcomeDialog.hardButton) {
				difficultyLevel = SudokuFactory.HARD;
			} else {
				System.exit(0);
			}
		
			game = SudokuFactory.createGame(difficultyLevel);
			
			gamePanel = new SudokuBoard(this, game);

			inGame = true;

			add(gamePanel, BorderLayout.CENTER);

			statusBar.startTime();

			pack();
				
		}
	}

	@Override
	public void actionPerformed(ActionEvent e) {
		Object source = e.getSource();

		if (source == newGameItem) {
			startNewGame();
		} else if (source == exitGameItem) {
			System.exit(0);
		} else if (source == aboutMenuItem) {
			JOptionPane.showMessageDialog(this, "About Sudoku:\n\n"
							+ "Another clone of the Japanese puzzle.\n\n"
							+ "Vegard Løkken (2012).\nVersion 0.25\n\n"
							+ "Feel free to share this game.");
		} else if (source == helpMenuItem) {
			JOptionPane.showMessageDialog(this, "The objective is to fill a "
							+ "9×9 grid with digits so that each column,\n"
							+ "each row, and each of the nine 3×3 sub-grids that "
							+ "compose the grid\n (also called \"boxes\", \"blocks\","
							+ "\"regions\", or \"sub-squares\")\n"
							+ "contains all of the digits from 1 to 9.");
		} else if (source == finishMenuItem) {
			int ans = JOptionPane.showConfirmDialog(this,
							"Are you sure you want me to finish the game? "
							+ "This can not be undone.", "Sure?",
							JOptionPane.YES_NO_OPTION);

			if (ans == JOptionPane.YES_OPTION) {
				gamePanel.finishGame();
			}
		} else if (source == retryGameItem) {
			gamePanel.retryGame();
		} else if (source == highscoreGameItem) {
			HighscoreService.showHighscoreTable(this, difficultyLevel);
		}
	}

	@Override
	public void itemStateChanged(ItemEvent e) {
		Object obj = e.getSource();
		if (obj instanceof CheckboxMenuItem) {

			CheckboxMenuItem source = (CheckboxMenuItem) e.getSource();

			if (source == lineAssistanceMenuItem) {
				lineAssistance = source.getState();
			} else if (source == numberAssistanceMenuItem) {
				numberAssistance = source.getState();
			} else if (source == redoMenuItem) {
				redo = source.getState();
			}
		}
	}

	public void gameOver() {
		inGame = false;

		HighscoreItem worst = HighscoreService.getWorst(difficultyLevel);

		if (HighscoreService.hasFreeSlots(difficultyLevel)
						|| statusBar.getTime() > worst.getTime()) {

			String name = JOptionPane.showInputDialog("Congratulations,\n"
							+ "you have completed the game and you've\nreached "
							+ "the highscore list.\n\nEnter your name:");
			while (name != null && name.length() > 30) {
				name = JOptionPane.showInputDialog("The name you entered "
								+ "is too long.\nPlease try again.");
			}

			if (name != null) {
				HighscoreService.addScore(name, difficultyLevel,
								statusBar.getTime() / 1000);

				HighscoreService.showHighscoreTable(this, difficultyLevel);
			}

			int ans = JOptionPane.showConfirmDialog(this,
							"Would you like to start a new game?",
							"New game?", JOptionPane.YES_NO_OPTION);

			if (ans == JOptionPane.YES_OPTION) {
				startNewGame();
			} else {
				System.exit(0);
			}
		} else {

			int ans = JOptionPane.showConfirmDialog(this,
							"Congratulations, you have completed the "
							+ "game. Would you like to start a new one?",
							"Congrats!", JOptionPane.YES_NO_OPTION);

			if (ans == JOptionPane.YES_OPTION) {
				startNewGame();
			} else {
				System.exit(0);
			}
		}
	}

	private WelcomeDialog showWelcomeDialog() {
		WelcomeDialog welcomeDialog = new WelcomeDialog();
		
		return welcomeDialog;
	}

	private class WelcomeDialog extends JDialog implements ActionListener {

		protected JButton easyButton, mediumButton, hardButton;
		protected JButton difficulty = null;

		public WelcomeDialog() {
			super(SudokuGame.this, "Welcome to Sudoku puzzle", true);

			easyButton = new JButton("Easy");
			easyButton.addActionListener(this);

			mediumButton = new JButton("Medium");
			mediumButton.addActionListener(this);

			hardButton = new JButton("Hard");
			hardButton.addActionListener(this);

			setLayout(new GridLayout(3, 1, 50, 5));

			add(new JLabel("Welcome to Sudoku puzzle!"));
			add(new JLabel("Please select a difficulty level:"));

			JPanel buttonPanel = new JPanel();
			buttonPanel.setLayout(new GridLayout(1, 3));
			buttonPanel.add(easyButton);
			buttonPanel.add(mediumButton);
			buttonPanel.add(hardButton);

			add(buttonPanel);

			pack();
			
			// Center
			setLocationRelativeTo(null);
			
			setVisible(true);
		}

		@Override
		public void actionPerformed(ActionEvent ae) {
			Object source = ae.getSource();

			if (source == easyButton || source == mediumButton ||
							source == hardButton) {
				difficulty = (JButton) source;
			}

			setVisible(false);
		}
		
	}

	public static void main(String[] args) {
		javax.swing.SwingUtilities.invokeLater(new Runnable() {

			@Override
			public void run() {
				new SudokuGame();
			}
		});
	}
}
